client/client.go

```diff
@@ -122,25 +123,41 @@ func bind(cfg *Config, conn ldaputil.Connection) error {
 		return errors.New("unable to bind due to lack of configured password")
 	}
 
-	if cfg.BindDN == "" {
-		return errors.New("must provide binddn")
-	}
-
-	origErr := conn.Bind(cfg.BindDN, cfg.BindPassword)
-	if origErr == nil {
+	if cfg.UPNDomain != "" {
+		origErr := conn.Bind(fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain), cfg.BindPassword)
+		if origErr == nil {
+			return nil
+		}
+		if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
+			return origErr
+		}
+		if err := conn.Bind(fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain), cfg.LastBindPassword); err != nil {
+			// Return the original error because it'll be more helpful for debugging.
+			return origErr
+		}
 		return nil
 	}
-	if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
-		return origErr
-	}
 
-	return conn.Bind(cfg.BindDN, cfg.LastBindPassword)
+	if cfg.BindDN != "" {
+		origErr := conn.Bind(cfg.BindDN, cfg.BindPassword)
+		if origErr == nil {
+			return nil
+		}
+		if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
+			return origErr
+		}
+		if err := conn.Bind(cfg.BindDN, cfg.LastBindPassword); err != nil {
+			// Return the original error because it'll be more helpful for debugging.
```

Comment:
I agree in most cases it is likely to be the most helpful error, but could we return both with some added context for maximum utility? e.g. "Attempted with current and previous password and got the following errors..."

Reply:
Good idea. Refactored this in https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/40/commits/d7d27a848321e1d1fd77395453c338546715a0fa. Both errors are now included along with a contextual prefix similar to your suggestion here.